src/keywrap/pkcs11/mod.rs

```diff
@@ -0,0 +1,333 @@
+            }
+        }
+
+        let plaintext = decrypt_pkcs11(&pkcs11_keys, annotation)?;
+
+        Ok(plaintext)
+    }
+
+
+    fn annotation_id(&self) -> &str {
+        "org.opencontainers.image.enc.keys.pkcs11"
+    }
+
+    fn no_possible_keys(&self,
+                        dcparameters: &HashMap<String, Vec<Vec<u8>>>
+    ) -> bool {
+        self.private_keys(dcparameters).is_none()
+    }
+
+    fn private_keys(&self,
+                    dcparameters: &HashMap<String, Vec<Vec<u8>>>,
+    ) -> Option<Vec<Vec<u8>>> {
+        dcparameters.get("pkcs11-yamls").cloned()
+    }
+
+    fn recipients(&self, _packet: String) -> Option<Vec<String>> {
+        Some(vec!["[pkcs11]".to_string()])
+    }
+}
+
+
+fn p11conf_from_params(dcparameters: &HashMap<String, Vec<Vec<u8>>>,
+) -> Result<Option<Pkcs11Config>> {
+    if dcparameters.contains_key("pkcs11-config") {
+        return Ok(Some(parse_pkcs11_config_file(&dcparameters["pkcs11-config"][0])?))
+    }
+    Ok(None)
+}
+
+fn add_pub_keys(dc: &DecryptConfig,
+                pubkeys: &Vec<Vec<u8>>,
+) -> Result<Vec<KeyType>> {
+    let mut pkcs11_keys = Vec::<KeyType>::new();
+    if pubkeys.is_empty() {
+        return Ok(pkcs11_keys);
+    }
+
+    let p11conf_opt = p11conf_from_params(&dc.param)?;
+
+    for pubkey in pubkeys {
+        let mut k = parse_public_key(pubkey, "PKCS11".to_string())?;
+        match &mut k {
+            KeyType::rpk(r) => {},
+            KeyType::pkfo(p) => {
+                if let Some(ref p11conf) = p11conf_opt {
+                    p.uriw.set_module_directories(&p11conf.module_directories);
+                    p.uriw.set_allowed_module_paths(&p11conf.allowed_module_paths);
+                }
+            },
+        }
+        pkcs11_keys.push(k);
+    }
+
+    Ok(pkcs11_keys)
+}
+
+
+
+
+#[cfg(test)]
+mod kw_tests {
+    use super::*;
+
+    //const SOFTHSM_SETUP: &str = "../../scripts/softhsm_setup";
+
+    #[test]
+    fn test_keywrap_pkcs11_success() {
+        let path_to_script = env!("CARGO_MANIFEST_DIR").to_string() + "/scripts/softhsm_setup";
+        let vs = create_valid_pkcs11_ccs().unwrap();
+        let valid_pkcs11_ccs = vs.0;
+        let shsm = vs.1;
+
+        // FIXME
+        //std::env::set_var("OCICRYPT_OAEP_HASHALG", "sha1");
+        std::env::set_var("OCICRYPT_OAEP_HASHALG", "sha256");
+
+        for cc in valid_pkcs11_ccs {
+            let kw = Pkcs11KeyWrapper{};
+
+            let data = "This is some secret text".as_bytes();
+
+            if let Some(ec) = cc.encrypt_config {
+                let wk = kw.wrap_keys(&ec, data).unwrap();
+                if let Some(dc) = cc.decrypt_config {
+                    let ud = kw.unwrap_keys(&dc, &wk).unwrap();
+                    assert_eq!(data, ud);
+                } else {
+                    assert!(false);
+                }
+            } else {
+                assert!(false);
+            }
+
+
+        }
+
+        shsm.run_softhsm_teardown(&path_to_script);
```

Comment:
The paired `setup` call actually comes from `create_valid_pkcs11_ccs`, which is confusing (and also led to me having to create two `path_to_script` variables. Fix this.